store/src/java/com/zimbra/cs/datasource/imap/RemoteFolder.java

```diff
@@ -71,10 +77,42 @@ public void delete() throws IOException {
 
     public RemoteFolder renameTo(String newName) throws IOException {
         info("renaming folder to '%s'", newName);
-        connection.rename(path, newName);
+
+        // ZBUG-2192
+        // execute only when last folder name's are different
+        // don't change any parent folder names
+        // using "path" as base name we are forming new folder name in such a way only last folder get's renamed at a time
+        
+        if (!isSameFolderName(newName)) {
+            String baseNewPath = path;
+            if (StringUtils.substringAfterLast(baseNewPath, PATH_SEPARATOR).isEmpty()
+                    && StringUtils.substringAfterLast(newName, PATH_SEPARATOR).isEmpty()) {
+                baseNewPath = newName;
+            } else {
+                baseNewPath = StringUtils.substringBeforeLast(baseNewPath, PATH_SEPARATOR) + PATH_SEPARATOR
+                        + StringUtils.substringAfterLast(newName, PATH_SEPARATOR);
+                if (StringUtils.endsWith(baseNewPath, PATH_SEPARATOR) || StringUtils.startsWith(baseNewPath, PATH_SEPARATOR)) {
+                    baseNewPath = StringUtils.removeEnd(baseNewPath, PATH_SEPARATOR);
+                    baseNewPath = StringUtils.removeStart(baseNewPath, PATH_SEPARATOR);
+                }
+            }
+            connection.rename(path, baseNewPath);
+        }
         return new RemoteFolder(connection, newName);
     }
 
+    // ZBUG-2192
+    // The helper method to make comparison of old and new folder name and
+    // returns boolean value
+    private boolean isSameFolderName(String newPathName) {
+
```

Comment:
remove empty line here and below, not needed

Reply:
Done